---
title: "DESeq2 Analysis with R: Description"
author: "Bioinfo-Core @ MPI-IE"
date: "`r date() `"
output:
  pdf_document:
    df_print: paged
urlcolor: blue
---

# Goals: 
The analysis of differential gene expression is a very common task, for which many 
advanced software packages have been developed. 
Using the DESeq2 package as example, we will introduce common analysis strategies to conduct gene expression analysis using R.

The following topics will be covered:

- Obtaining and loading expression data
- Understanding data, metadata and designs
- Data exploration and quality controls
- Data filtering and normalisation
- DESeq2 workflow
- Defining contrasts, understanding coefficients and calling differential genes
- Data visualisation
- Documenting your reproducible analysis using notebooks

In case you do have your own (multifactorial) RNA-seq data, we encourage you to bring this data to the course. It should be in the form of a count matrix. Don't hesitate to get in touch with the bioinfo-core in case you are not sure what this entails or need help generating this matrix.

# Requirements:

* **basic R understanding**: Introductory material has been covered by our introductory [course](https://github.com/maxplanck-ie/Rintro). We will spend a (very) limited amount of time refreshing some basic concepts, so we encourage you to re-familiarise yourself with the Rstudio IDE and [data types and structures in R](https://swcarpentry.github.io/r-novice-inflammation/13-supp-data-structures/) if that's needed.

* an **account for [workbench](https://workbench.ie-freiburg.mpg.de)**: We will use the Rstudio server at the MPI-IE. If you do not have an account yet, you will need to request an account from core-IT. Please do so well before the course starts. You can use a local installation of R as well. Note that we will be using R 4.1.3 during this course.

* **Know your keyboard!** Special characters will be needed throughout the course, e.g. $ , | , [, ], {, }, >, <, #, ~, &, ^, %, !, ?

***



