---
title: "deseq"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# add libraries
library("DESeq2")

# Read count matrix
countdata <- read.table("/data/.../counts.tsv", header=TRUE, check.names = TRUE)

# Read smaplesheet , detects conditions, generate the formula
sampleInfo <- read.table("/data/.../samplehseet.tsv", header=TRUE, check.names = TRUE, stringsAsFactor = F)
sampleInfo$genotype <- as.factor(sampleInfo$genotype)
sampleInfo$treatment <- as.factor(sampleInfo$treatment)

# Different steps in a differential expression analysis:

# build the matrix
d<-as.formula(~condition + treatment + condition:treatment)
dds <- DESeq2::DESeqDataSetFromMatrix(countData = countdata, colData = sampleInfo, design =d)

# deseq analysis
#dds <- DESeq(dds)

#estimate size factor
# DESeq2 will automatically estimate the size factors when performing the differential expression analysis. However, if estimateSizeFactors() is called, then DESeq2 will use these values.
# To normalize the count data, DESeq2 calculates size factors for each sample using the "median of ratios" method.

dds <- estimateSizeFactors(dds)
sizeFactors(dds)

# alternative1 deseq analysis
# dds <- DESeq(dds)

# Find the relation between the size factors and the values of the columns:
colSums(counts(dds))

#the larger size factors correspond to the samples with higher sequencing depth. To generate the normalized counts we need to divide the counts by the size factors. 
# This accounts for the differences in sequencing depth between samples.

# compare the raw sum with the normalised ones:
colSums(counts(dds, normalized=T))

# The importance of dispersion during differential expression analysis

# gene-wise dispersion
dds <- estimateDispersions(dds)

#visualise dispersion
plotDispEsts(dds)


#steps required to generate a results table for pairwise comparisons (Wald test)

# make a pairwise comparison (using contrast)
resultsNames(dds) # to see what names to use
contrast <- list(resultsNames(dds)[1], resultsNames(dds)[2])

#If there is only two factor levels, There is no need to  specify contrasts,
#DESeq2 will choose the base factor level based on alphabetical order of the levels.

res <- results(dds, contrast = contrast, alpha = 0.05)

# What is in the res?
res %>% 
data.frame() %>% 
View()

# to get the coef name check resultsNames(dds)
res_shrink <- lfcShrink(dds, coef="...", type="apeglm")


#MA plot

plotMA(res, ylim=c(-2,2))
plotMA(res_shrink, ylim=c(-2,2))


# summary
summary(res_tableOE, alpha = 0.05)
