---
title: "DESeq2 Analysis with R: Part 00"
author: "Bioinfo-Core @ MPI-IE"
#date: "`r date() `"
output: html_document
---

# Goals: 

<!-- ![Typical examples from R.](../images/DESeq2_workflow.png) -->

After the course you will be able to

* understand bioinformatics workflows: data input $\to$ output
* perform and understand quality controls
* visualize data and results
* apply complex experimental designs
* obtain differential genes
* share workflows with notebook

***

# Prerequistites
* R introdcution: data formats, functions, help, plotting, work with packages
* account for rstudio.ie-freiburg.mpg.de
* Patience

***

# Credits
There is lots of useful material out there. This course is heavily influenced by those tutorials

 * https://hbctraining.github.io/DGE_workshop_salmon_online/
 * http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

***

# Overview

## Part I: Getting and understanding the data

* Warmup & R-epetition
* Data, Metadata & Design
* Experimental Design
* Data Exploration
* Quality Control


## Part II: Data Transformations and Model

* Quality Controls & Filtering
* Data Transformations and Normalization
* Size Factors
* Modeling Count Data: NB and Over-dispersion
* Running DESeq2

## Part III: Differentially Expressed Genes

* Multifactorial Designs & Complex Contrasts
* Hypothesis tests: Wald test & LRT
* Inspection & Visualization
* Assessing Overlaps
* Gene Annotations
* Exporting results

## Part IV: On your own 
* Run a full DESeq2 Analysis

***



