---
title: "deseq2"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_depth: 2
    toc_float: true
    df_print: paged
---
# Goals:
* Different steps of a differential gene expression analysis
  * multi-factor design
  * Normalisation: Estimating size factors
  * Checking for the dispertion
  * Building up contrasts
  * Make a results table
  * Summarize and extract a list of significant differentially expressed genes
  * Visualize the results
  * LRT vs. Wald test
  
```{r installation, eval=FALSE, message=FALSE, warning=FALSE, paged.print=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("DESeq2")
BiocManager::install("apeglm")
install.packages("tidyverse")
install.packages("dplyr")
install.packages("ggrepel")
```
# add libraries
```{r add libraries, message=FALSE, warning=FALSE, paged.print=FALSE}
library("DESeq2")
library("dplyr")
library("ggplot2")
library("ggrepel")
```
# Get the data directroy
```{r Get the data directroy, message=FALSE, warning=FALSE}
setwd("/data/processing1/leily/r_course/Rdeseq2/")
data_dir <- paste(getwd(), "data", "qc", sep = .Platform$file.sep)
#print(data_dir)

# Read count matrix
countdata <- read.table(paste(data_dir,"counts_qc.tsv", sep = "/"), header=TRUE, check.names = TRUE)
#head(countdata)

# Read smaplesheet , detects conditions, generate the formula
sampleInfo <- read.table(paste(data_dir,"counts_qc_meta.tsv", sep = "/"), header=TRUE, check.names = TRUE, stringsAsFactor = F)
sampleInfo$cellType <- as.factor(sampleInfo$cellType)
sampleInfo$txType <- as.factor(sampleInfo$txType)
head(sampleInfo)
```

# Building a deseq data set with a multi-factor design:
The formula for the model includes all sources of variation in the data:
cellType, txType and the difference as the effect of cellType on txType (cellType:txType)

```{r build the matrix, message=FALSE, warning=FALSE}
d<-as.formula(~cellType + txType + cellType:txType)
dds <- DESeq2::DESeqDataSetFromMatrix(countData = countdata, colData = sampleInfo, design =d)
```

# Estimate size factor
DESeq2 automatically estimates the size factors when performing the differential expression analysis. However, if `estimateSizeFactors()` is called, then DESeq2 will use these values.

To normalize the count data, DESeq2 calculates size factors for each sample using the "median of ratios" method.
```{r estimate size factors, message=FALSE, warning=FALSE}
dds <- estimateSizeFactors(dds)
sizeFactors(dds)
```

# Find the relation between the size factors and the values of the columns:
```{r colsums, message=FALSE, warning=FALSE}
colSums(counts(dds))
```
Larger size factors correspond to the samples with higher sequencing depth. To generate the normalized counts we need to divide the counts by the size factors. This accounts for the differences in sequencing depth between samples.

# compare the raw sum with the normalised ones:
```{r norm colsums, message=FALSE, warning=FALSE}
colSums(counts(dds, normalized=T))
```

# Importance of dispersion during differential expression analysis
DESeq2 uses a measure of variation called dispersion, which accounts for a gene’s variance and mean expression level. Dispersion is calculated by `Var = μ + α*μ^2`, where `α` = dispersion, `Var` = variance, and `μ` = mean
# Estimate dispersion
```{r estimate dispersions, message=FALSE, warning=FALSE}
# To generate more accurate estimates of variation based on the mean expression level of the gene using a method called ‘shrinkage’. DESeq2 assumes that genes with similar expression levels should have similar dispersion. (??)
dds <- estimateDispersions(dds)
```
# Visualise dispersion
```{r Visualise dispersions, message=FALSE, warning=FALSE}
plotDispEsts(dds)
```
# Run deseq analysis
```{r deseq analysis, message=FALSE, warning=FALSE}
dds <- DESeq(dds)
```
#steps required to generate a results table for pairwise comparisons (Wald test)
text
# make a pairwise comparison (using contrast)
Contrasts can be given as a list of 2 character vectors:
the names of the fold changes for the level of ineterest, and the names of the fold changes for the base level.
```{r contrast, message=FALSE, warning=FALSE}
resultsNames(dds) # to see what names to use
# contrast_oe <- c("cellType", "JJ", "BM")
contrast <- list(resultsNames(dds)[1], resultsNames(dds)[2])
```
If there is only two factor levels, There is no need to  specify contrasts, DESeq2 will choose the base factor level based on the alphabetical order of the levels.

By default DESeq2 uses the Wald test to identify genes that are differentially expressed between two sample classes
```{r result, message=FALSE, warning=FALSE}
# which contrast are we interested in?
res <- results(dds, contrast = contrast, alpha = 0.05)
```
# What is in the res?
```{r view res, message=FALSE, warning=FALSE}
res %>% data.frame() %>% head()
```
* baseMean: mean of normalized counts for all samples
* log2FoldChange: log2 fold change
* lfcSE: standard error
* stat: Wald statistic
* pvalue: Wald test p-value
* padj: BH adjusted p-values

TODO: Add a plot of mean normalized counts vs log10(p values)

To generate more accurate log2 foldchange (LFC) estimates, DESeq2 allows for the shrinkage of the LFC estimates toward zero when the information for a gene is low, which could include:
- Low counts
- High dispersion values
LFC shrinkage uses information from all genes to generate more accurate estimates. Specifically, the distribution of LFC estimates for all genes is used (as a prior) to shrink the LFC estimates of genes with little information or high dispersion toward more likely (lower) LFC estimates.
```{r shrink, message=FALSE, warning=FALSE}
# which coef to use?
resultsNames(dds)
res_shrink <- lfcShrink(dds, coef="cellType_JJ_vs_BM", type="apeglm")
```
To get the coef name check resultsNames(dds)

text
#MA plot
```{r MA plot, message=FALSE, warning=FALSE}
plotMA(res, ylim=c(-2,2))
plotMA(res_shrink, ylim=c(-2,2))
```
text
# summary
```{r summary, message=FALSE, warning=FALSE}
summary(res_shrink, alpha = 0.05)
cutoff <- 0.05
res_sig <- subset(res_shrink, padj<cutoff)
res_sig_sorted = res_sig[order(res_sig$padj), ]
head(res_sig_sorted)
```
# Visualisation
```{r visualisation, message=FALSE, warning=FALSE}
# plot a single gene
plotCounts(dds, gene = "ENSG00000164062", intgroup = "cellType")
# To see the data underneath the figure
d <- plotCounts(dds, gene = "ENSG00000164062", intgroup = "cellType", returnData=TRUE)
d %>% View()
# can be used for ggplot
ggplot(d, aes(x = cellType, y = count, color = cellType)) + 
    geom_point(position=position_jitter(w = 0.1,h = 0)) +
    geom_text_repel(aes(label = rownames(d))) + 
    theme_bw() +
    ggtitle("ENSG00000164062") +
    theme(plot.title = element_text(hjust = 0.5))
```
```{r heatmap, message=FALSE, warning=FALSE}


```

#Likelihood ratio test
```{r lrt, message=FALSE, warning=FALSE}


```

