---
title: "DESeq2 Analysis with R: Part 00"
author: "Bioinfo-Core @ MPI-IE"
#date: "`r date() `"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_depth: 1
    toc_float: true
    df_print: paged
---

# Goals

<!-- ![DESeq workflow](../images/workflow_DESeq2.png) -->

After the course you will be able to

* perform a full bioinformatics workflow (DESeq2): data input $\to$ output  
* perform exploratory analysis: quality controls, filters  
* visualize data and results with modern tools  
* apply complex experimental designs  
* obtain and visualise differential genes from simple or complex comparison setups  
* share workflows with notebook  

***

# Prerequistites

* R introduction: data formats, functions, help, plotting, work with packages
* (linux) account for https://workbench.ie-freiburg.mpg.de or your own local R/Rstudio installation.
* Patience

Note that this course will be based on R version 4.1.3, Bioconductor version 3.14.

***

# Credits

There is lots of useful material out there. This course is heavily influenced by those tutorials

 * https://hbctraining.github.io/DGE_workshop_salmon_online/
 * http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
 
We will also be using publically available data, which has been altered for educational purposes (and should thus be treated as such).

***

# Overview

## 1. Part I: Getting and understanding the data

* Warmup & R-epetition
* Data, Metadata & Design
* Data Exploration
* Quality Control


## 2. Part II: Data Transformations and Model

* Quality Controls & Filtering
* Data Transformations and Normalization
* Size Factors
* Modeling Count Data: NB and Over-dispersion
* Running DESeq2

## 3. Part III: Differentially Expressed Genes

* Multifactorial Designs & Complex Contrasts
* Hypothesis tests: Wald test & LRT
* Inspection & Visualization
* Gene Annotations
* Exporting results

## 4. Part IV: On your own 
* Run a full multifactorial DESeq2 Analysis. We'll be around to help.

***

# Libraries

During the course we'll use a (limited) number of libraries. These are listed below and are considered installed in the rest of the course.
We make use of the custom 'my_load_namespaces' function (Thanks Adrian!) to make sure our libraries are installed and available.


```{r}
my_load_namespaces <- function(packagesToLoad=c(), coreLibraries=c()) {
  # coreLibraries = not explicitly loaded, but required anyway (e.g. dependencies)
  installed_pkgs <- installed.packages()
  
  if (! "BiocManager" %in% installed_pkgs) install.packages("BiocManager")
  
  lapply(c(coreLibraries, packagesToLoad),
         function(x) if (! x %in% installed_pkgs) BiocManager::install(x, ask = FALSE))
  
  lapply(packagesToLoad, function(pkg) {
    suppressPackageStartupMessages(require(pkg, character.only = TRUE))
  })
  
  TRUE
}
```


```{r Libraries, echo=FALSE,warning=FALSE,message=FALSE }
my_load_namespaces(c("DESeq2", "tidyverse"), c())
```

***


