<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Bioinfo-Core @ MPI-IE" />


<title>DESeq2 Analysis with R: Part 02</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Rdeseq2 - RNA-seq analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    
  </a>
</li>
<li>
  <a href="outline.html">
    <span class="fa fa-check"></span>
     
    outline
  </a>
</li>
<li>
  <a href="day01.html">
    <span class="fa fa-battery-quarter"></span>
     
    Day 1
  </a>
</li>
<li>
  <a href="day02.html">
    <span class="fa fa-battery-half"></span>
     
    Day 2
  </a>
</li>
<li>
  <a href="day03.html">
    <span class="fa fa-battery-three-quarters"></span>
     
    Day 3
  </a>
</li>
<li>
  <a href="day04.html">
    <span class="fa fa-battery-full"></span>
     
    Day 4
  </a>
</li>
<li>
  <a href="summary.html">
    <span class="fa fa-truck-medical"></span>
     
    summary
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">DESeq2 Analysis with R: Part 02</h1>
<h4 class="author">Bioinfo-Core @ MPI-IE</h4>
<h4 class="date">Fri Apr 28 09:43:51 2023</h4>

</div>


<div id="goals-deseq2-workflow" class="section level2">
<h2>Goals: DESeq2 workflow</h2>
<ul>
<li>repeat: generating DESeq-object (data, metadata, design)</li>
<li>filtering: samples and genes</li>
<li>transformations: normalization and sizefactors</li>
<li>modeling: variance + dispersion</li>
<li>statistical tests</li>
<li>extracting results</li>
</ul>
</div>
<div id="preparation-repeat" class="section level2">
<h2>Preparation (Repeat)</h2>
<div id="load-libraries" class="section level3">
<h3>Load Libraries</h3>
<pre class="r"><code>suppressPackageStartupMessages({
library(&quot;tidyverse&quot;)
library(&quot;DESeq2&quot;)
library(&quot;pheatmap&quot;)
})</code></pre>
</div>
<div id="load-data" class="section level3">
<h3>Load data</h3>
<pre class="r"><code>data &lt;- readr::read_tsv(&quot;data/myeloma/myeloma_counts.tsv&quot;)</code></pre>
<pre><code>## Rows: 57905 Columns: 19
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr  (1): gene_id
## dbl (18): JJ_CTRL_1, BM_CTRL_2, BM_CTRL_3, BM_AMIL_1, BM_AMIL_2, BM_AMIL_3, ...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<pre class="r"><code>data &lt;- data %&gt;% 
  column_to_rownames(var=&quot;gene_id&quot;) %&gt;%
  mutate(across(where(is.double), as.integer))

data %&gt;% str()</code></pre>
<pre><code>## &#39;data.frame&#39;:    57905 obs. of  18 variables:
##  $ JJ_CTRL_1: int  0 0 1637 453 719 2 2 61 2354 3025 ...
##  $ BM_CTRL_2: int  0 0 1792 614 961 3 1 76 2808 4071 ...
##  $ BM_CTRL_3: int  0 0 2368 854 1280 6 2 98 3963 4837 ...
##  $ BM_AMIL_1: int  0 0 2365 623 937 0 3 180 1298 2242 ...
##  $ BM_AMIL_2: int  0 0 4808 585 1247 0 0 176 1240 2818 ...
##  $ BM_AMIL_3: int  0 0 2098 431 852 0 0 163 812 1981 ...
##  $ BM_CTRL_1: int  0 0 1664 687 884 1 2498 1543 1590 2097 ...
##  $ JJ_CTRL_2: int  0 0 1878 869 1058 4 2860 2165 2018 2906 ...
##  $ JJ_CTRL_3: int  0 0 1367 619 616 1 1991 1513 1331 1971 ...
##  $ JJ_AMIL_1: int  0 0 1155 512 840 3 2274 1262 552 1178 ...
##  $ JJ_AMIL_2: int  0 0 3020 953 1320 3 3197 2265 1316 2099 ...
##  $ JJ_AMIL_3: int  0 0 3834 1049 1161 5 3817 2411 1194 2075 ...
##  $ BM_TG_1  : int  0 0 2461 845 1944 5 3 53 3384 3274 ...
##  $ BM_TG_2  : int  0 0 1962 757 1528 1 0 40 2918 2785 ...
##  $ BM_TG_3  : int  0 0 2417 859 2106 3 1 43 3516 3610 ...
##  $ JJ_TG_1  : int  0 0 1601 823 1350 2 2917 1280 1999 1889 ...
##  $ JJ_TG_2  : int  0 0 3110 1342 2569 2 4345 1998 3321 3467 ...
##  $ JJ_TG_3  : int  0 0 2129 904 1654 3 3278 1434 2352 2404 ...</code></pre>
<pre class="r"><code>data %&gt;% dim()</code></pre>
<pre><code>## [1] 57905    18</code></pre>
<pre class="r"><code>data %&gt;% head()</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["JJ_CTRL_1"],"name":[1],"type":["int"],"align":["right"]},{"label":["BM_CTRL_2"],"name":[2],"type":["int"],"align":["right"]},{"label":["BM_CTRL_3"],"name":[3],"type":["int"],"align":["right"]},{"label":["BM_AMIL_1"],"name":[4],"type":["int"],"align":["right"]},{"label":["BM_AMIL_2"],"name":[5],"type":["int"],"align":["right"]},{"label":["BM_AMIL_3"],"name":[6],"type":["int"],"align":["right"]},{"label":["BM_CTRL_1"],"name":[7],"type":["int"],"align":["right"]},{"label":["JJ_CTRL_2"],"name":[8],"type":["int"],"align":["right"]},{"label":["JJ_CTRL_3"],"name":[9],"type":["int"],"align":["right"]},{"label":["JJ_AMIL_1"],"name":[10],"type":["int"],"align":["right"]},{"label":["JJ_AMIL_2"],"name":[11],"type":["int"],"align":["right"]},{"label":["JJ_AMIL_3"],"name":[12],"type":["int"],"align":["right"]},{"label":["BM_TG_1"],"name":[13],"type":["int"],"align":["right"]},{"label":["BM_TG_2"],"name":[14],"type":["int"],"align":["right"]},{"label":["BM_TG_3"],"name":[15],"type":["int"],"align":["right"]},{"label":["JJ_TG_1"],"name":[16],"type":["int"],"align":["right"]},{"label":["JJ_TG_2"],"name":[17],"type":["int"],"align":["right"]},{"label":["JJ_TG_3"],"name":[18],"type":["int"],"align":["right"]}],"data":[{"1":"0","2":"0","3":"0","4":"0","5":"0","6":"0","7":"0","8":"0","9":"0","10":"0","11":"0","12":"0","13":"0","14":"0","15":"0","16":"0","17":"0","18":"0","_rn_":"ENSG00000000003"},{"1":"0","2":"0","3":"0","4":"0","5":"0","6":"0","7":"0","8":"0","9":"0","10":"0","11":"0","12":"0","13":"0","14":"0","15":"0","16":"0","17":"0","18":"0","_rn_":"ENSG00000000005"},{"1":"1637","2":"1792","3":"2368","4":"2365","5":"4808","6":"2098","7":"1664","8":"1878","9":"1367","10":"1155","11":"3020","12":"3834","13":"2461","14":"1962","15":"2417","16":"1601","17":"3110","18":"2129","_rn_":"ENSG00000000419"},{"1":"453","2":"614","3":"854","4":"623","5":"585","6":"431","7":"687","8":"869","9":"619","10":"512","11":"953","12":"1049","13":"845","14":"757","15":"859","16":"823","17":"1342","18":"904","_rn_":"ENSG00000000457"},{"1":"719","2":"961","3":"1280","4":"937","5":"1247","6":"852","7":"884","8":"1058","9":"616","10":"840","11":"1320","12":"1161","13":"1944","14":"1528","15":"2106","16":"1350","17":"2569","18":"1654","_rn_":"ENSG00000000460"},{"1":"2","2":"3","3":"6","4":"0","5":"0","6":"0","7":"1","8":"4","9":"1","10":"3","11":"3","12":"5","13":"5","14":"1","15":"3","16":"2","17":"2","18":"3","_rn_":"ENSG00000000938"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="load-metadata" class="section level3">
<h3>Load Metadata</h3>
<pre class="r"><code>metadata &lt;- readr::read_tsv(&quot;data/myeloma/myeloma_meta.tsv&quot;)</code></pre>
<pre><code>## Rows: 18 Columns: 3
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (3): sample, celltype, condition
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<pre class="r"><code>metadata &lt;- metadata %&gt;% 
  mutate(celltype = factor(celltype)) %&gt;%
  mutate(condition = factor(condition)) %&gt;%
  column_to_rownames(&quot;sample&quot;)  

metadata %&gt;% str()</code></pre>
<pre><code>## &#39;data.frame&#39;:    18 obs. of  2 variables:
##  $ celltype : Factor w/ 2 levels &quot;BM&quot;,&quot;JJ&quot;: 2 1 1 1 1 1 1 2 2 2 ...
##  $ condition: Factor w/ 3 levels &quot;AMIL&quot;,&quot;DMSO&quot;,..: 2 2 2 1 1 1 2 2 2 1 ...</code></pre>
<pre class="r"><code>metadata %&gt;% dim()</code></pre>
<pre><code>## [1] 18  2</code></pre>
<pre class="r"><code>metadata %&gt;% head()</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["celltype"],"name":[1],"type":["fct"],"align":["left"]},{"label":["condition"],"name":[2],"type":["fct"],"align":["left"]}],"data":[{"1":"JJ","2":"DMSO","_rn_":"JJ_CTRL_1"},{"1":"BM","2":"DMSO","_rn_":"BM_CTRL_2"},{"1":"BM","2":"DMSO","_rn_":"BM_CTRL_3"},{"1":"BM","2":"AMIL","_rn_":"BM_AMIL_1"},{"1":"BM","2":"AMIL","_rn_":"BM_AMIL_2"},{"1":"BM","2":"AMIL","_rn_":"BM_AMIL_3"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<pre class="r"><code>#check if the col and rows are in the same order:
all(rownames(metadata) == colnames(data))</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code># Convert to factors
metadata$celltype &lt;- metadata$celltype %&gt;% as.factor()
metadata$condition &lt;- metadata$condition %&gt;% as.factor()</code></pre>
</div>
<div id="create-design" class="section level3">
<h3>Create Design</h3>
<pre class="r"><code>my_design &lt;- ~celltype + condition</code></pre>
</div>
<div id="create-deseq-object" class="section level3">
<h3>Create DESeq Object</h3>
<pre class="r"><code>dds &lt;- DESeqDataSetFromMatrix(countData=data, colData=metadata, design=my_design)

class(dds)</code></pre>
<pre><code>## [1] &quot;DESeqDataSet&quot;
## attr(,&quot;package&quot;)
## [1] &quot;DESeq2&quot;</code></pre>
<pre class="r"><code>dim(dds)</code></pre>
<pre><code>## [1] 57905    18</code></pre>
</div>
<div id="first-inspection-pca" class="section level3">
<h3>First Inspection (PCA)</h3>
<p>Previously you noticed that the PCA shows a problem with our
data.</p>
<pre class="r"><code># define function (see day 1)
myPlotPCA = function(dds, nt=500) {
  M &lt;- counts(dds)                 # get counts from dds
  Mt &lt;- log2(M + 1)                # a simple (and fast) transformation
  ann &lt;- data.frame(colData(dds))  # get annotations from colData = metadata
  cn_col &lt;- c(&quot;#1B9E77&quot;, &quot;#D95F02&quot;, &quot;#7570B3&quot;)   # condition colours
  names(cn_col) &lt;- levels(ann$condition) # name the color vector

  top &lt;- Mt %&gt;% rowVars() %&gt;% order(decreasing=TRUE) %&gt;% head(nt)  # calculate variance for each row and sort
  pca &lt;- Mt[top,] %&gt;% t() %&gt;% prcomp(scale=TRUE)                    # perform PCA

  rld_PCA &lt;- as.data.frame(pca$x[,1:2])                     # only take PC1 and PC2 (column 1 + 2)
  rld_PCA$condition &lt;- ann[rownames(rld_PCA),&quot;condition&quot;]   # add condition label from ann
  rld_PCA$celltype  &lt;- ann[rownames(rld_PCA),&quot;celltype&quot;]    # add celltype label from ann

  ggplot(rld_PCA, aes(PC1, PC2, color=condition, shape=celltype)) +
    geom_point(size=3) +
    scale_colour_manual(values=cn_col)
}


myPlotPCA(dds)</code></pre>
<p><img src="day02_files/figure-html/old_PCA-1.png" width="80%" style="display: block; margin: auto;" /></p>
<blockquote>
<p><strong>Poll 2.1:</strong> What should we do next ?</p>
</blockquote>
</div>
</div>
<div id="filtering" class="section level2">
<h2>Filtering</h2>
<div id="filter-samples---10-min" class="section level3">
<h3>Filter Samples - 10 min</h3>
<p><strong>Task</strong>: Filter the problematic samples (columns 1 and
7) from dds object and produce a new PCA plot.</p>
<p>Hint: Although the <code>dds</code> object is a more complex data
class it also has dimensions assigned – check <code>dim(dds)</code> –
and samples can be un/selected like columns in a standard
dataframes.</p>
<pre class="r"><code>dds_clean &lt;- dds[, -c(1,7)] 
myPlotPCA(dds_clean)</code></pre>
<p><img src="day02_files/figure-html/filter_samples-1.png" width="80%" style="display: block; margin: auto;" /></p>
</div>
<div id="filter-genes---10-min" class="section level3">
<h3>Filter Genes - 10 min</h3>
<ul>
<li>remove genes (rows) that have very low counts</li>
<li>arbitrary threshold: &lt; 1, &lt; 10</li>
<li>benefits:
<ul>
<li>increase robustness of analysis</li>
<li>reduces memory size of ‘dds’ object =&gt; increases the speed</li>
</ul></li>
</ul>
<p><strong>Task</strong> Filter out the genes that have a total amount
of only zero or 1 counts.</p>
<p>Hint: Have a look at the output of
<code>dds_clean %&gt;% counts %&gt;% rowSums() %&gt;% head()</code></p>
<pre class="r"><code>rs = dds_clean %&gt;% counts %&gt;% rowSums()
dds_clean = dds_clean[rs&gt;1,]</code></pre>
<blockquote>
<p><strong>Poll 2.2</strong> How many samples and genes are left in the
filtered data set?</p>
</blockquote>
</div>
</div>
<div id="relevel-factors---10-min" class="section level2">
<h2>Relevel Factors - 10 min</h2>
<p><strong>Remember</strong>: factors are used to represent categorical
data !</p>
<p>Internally they are stored as integers and per default they are
sorted <em>alphanumerically</em>. It may be more convenient to sort them
by some user defined order - this is called relevelling</p>
<center>
<div class="figure">
<img src="images/day2_factor_order_comparison.png" style="width:60.0%"
alt="" />
<p class="caption">Factor order</p>
</div>
</center>
<p><br></p>
<blockquote>
<p><strong>Task</strong> Reorder the condition with ‘DMSO’ as baseline -
2 min.</p>
</blockquote>
<p>Hint: Use the <code>relevel</code> function</p>
<p>This is the order of the factors before and after re-arrangement:</p>
<pre class="r"><code>#initial factor arrangement:
dds_clean$condition</code></pre>
<pre><code>##  [1] DMSO DMSO AMIL AMIL AMIL DMSO DMSO AMIL AMIL AMIL TG   TG   TG   TG   TG  
## [16] TG  
## Levels: AMIL DMSO TG</code></pre>
<pre class="r"><code>dds_clean$condition &lt;- relevel(dds_clean$condition, ref=&#39;DMSO&#39;)
#dds_clean$condition &lt;- factor(dds_clean$condition, levels = c (&#39;DMSO&#39;, &#39;TG&#39;, &#39;AMIL&#39;)) # alternatively

dds_clean$condition</code></pre>
<pre><code>##  [1] DMSO DMSO AMIL AMIL AMIL DMSO DMSO AMIL AMIL AMIL TG   TG   TG   TG   TG  
## [16] TG  
## Levels: DMSO AMIL TG</code></pre>
<p><strong>Notice:</strong> Per default DESeq2 will make comparisons
with respect to the 1st level, unless specified more explicitly
(contrasts, see tomorrow).</p>
</div>
<div id="preparation-summary" class="section level2">
<h2>Preparation Summary</h2>
<p>We have now covered the most important preparation steps:</p>
<ul>
<li>load library and data</li>
<li>define DESeq data object
<ul>
<li>count data: <code>counts()</code> – not normalized. data frame of
integers</li>
<li>meta data: <code>colData()</code> – added data on samples. data
frame of numerical or categorical variables</li>
<li>design: <code>design()</code> – variables to be modeled. formula
will be converted to design matrix with dummy variables</li>
</ul></li>
<li>Filtering &amp; Transformation &amp; Relevel</li>
</ul>
<blockquote>
<p>This is <strong>before</strong> any statistical analysis ! A
real-world analysis would likely be iterative</p>
</blockquote>
<blockquote>
<p><strong>Message</strong>: Document your decisions, parameters,
workflows and software versions carefully - use notebooks !</p>
</blockquote>
</div>
<div id="break---10-min" class="section level2">
<h2>Break - 10 min</h2>
<hr />
</div>
<div id="the-deseq-workflow-60-min" class="section level2">
<h2>The DESeq workflow (60 min)</h2>
<p>Everything we did until now was before any statistical analysis.
Quite some work stil needs to happen. Luckily, our life is made easy
(again), as DESeq2 wrap up the rest in one convenient function:</p>
<blockquote>
<p>dds &lt;- DESeq(dds)</p>
</blockquote>
<pre class="r"><code>dds_clean &lt;- DESeq(dds_clean)</code></pre>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<p>There are several steps to generating differential expression
results, and they are wrapped together in the <strong>DESeq()</strong>
function. The 3 main steps are of the workflow are:</p>
<ul>
<li>Normalization - <em>estimateSizeFactors()</em></li>
<li>estimating dispersions - <em>estimateDispersions()</em></li>
<li>general linear model fit for each gene and testing -
<em>nbinomWaldTest()</em></li>
</ul>
<div id="a.-normalization" class="section level3">
<h3>a. Normalization</h3>
<p>As mentioned yesterday, raw counts cannot be directly used to compare
samples. To allow comparison we need to normalize. Note that this is
also needed for exploratory analysis and visualisation.</p>
<p>A couple of reasons for normalization include:</p>
<hr />
<p><em>sequencing depth</em></p>
<div class="figure">
<img src="images/day2_normalization_seqdepth.png" style="width:50.0%"
alt="" />
<p class="caption">‘doubling of expression due to sequencing depth’.
Image from github.com/hbctraining</p>
</div>
<hr />
<p><em>gene length</em></p>
<div class="figure">
<img src="images/day2_normalization_genelength.png" style="width:35.0%"
alt="" />
<p class="caption">‘gene x and y have similar levels of expression,
though gene x has way more reads aligned due to its length.’. Image from
github.com/hbctraining</p>
</div>
<hr />
<p><em>compositional bias</em></p>
<div class="figure">
<img src="images/day2_normalization_compbias.png" style="width:50.0%"
alt="" />
<p class="caption">‘compositional bias through contaminant (e.g. rRNA)’.
Image from github.com/hbctraining</p>
</div>
<hr />
<p>A number of commonly used normalization techniques:</p>
<table>
<thead>
<tr class="header">
<th>method</th>
<th align="left">factors taking into account</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>CPM</td>
<td align="left">sequencing depth</td>
</tr>
<tr class="even">
<td>RPKM</td>
<td align="left">sequencing depth and gene length</td>
</tr>
<tr class="odd">
<td>DESeq2 median of ratios</td>
<td align="left">sequencing depth and RNA composition</td>
</tr>
</tbody>
</table>
<p>Workflow for the ‘median of ratios’:</p>
<ul>
<li>Take geometric mean per gene (of all samples!) -&gt;
‘pseudosample’</li>
<li>calculate the count / pseudosample ratio</li>
<li>‘sizefactor’ == median(ratios)</li>
</ul>
<p>and many others exist.</p>
<p>Remember that DESeq2 needs <strong>raw counts</strong> as input (upon
creating the dds object).</p>
<blockquote>
<p><strong>Task </strong>: Plot a heatmap of the first 10 rows of your
counts (non-normalized), deseq normalized counts and if you are up for
some manual labour, sequencing depth normalized counts.</p>
</blockquote>
<p>Hint, counts() can return DESeq normalized counts as well with the
argument normalized=TRUE! Hint2: geometric mean of a vector x can be
calculated with exp(mean(log(x)))</p>
<pre class="r"><code>ann &lt;- data.frame(colData(dds_clean))
ann$seq_depth &lt;- counts(dds_clean) %&gt;% colSums()/1e6    # sequencing depth in millions 

pheatmap(
  counts(dds_clean) %&gt;% head(10),
  main=&quot;Raw&quot;,
  cluster_cols=FALSE,
  cluster_rows=FALSE,
  annotation=ann
)</code></pre>
<p><img src="day02_files/figure-html/heatmap_raw-1.png" width="672" /></p>
<pre class="r"><code>pheatmap(
  counts(dds_clean, normalized=TRUE) %&gt;% head(10),
  main=&quot;median of ratios&quot;,
  cluster_cols=FALSE,
  cluster_rows=FALSE,
  annotation=ann
)</code></pre>
<p><img src="day02_files/figure-html/heatmap%20deseq_normalized-1.png" width="672" /></p>
<pre class="r"><code>sfs &lt;- (counts(dds_clean) %&gt;% colSums()) / ((counts(dds_clean) %&gt;% colSums()) %&gt;% log() %&gt;% mean() %&gt;% exp())
ann$sfs &lt;- sfs
pheatmap(
  sweep(counts(dds_clean), 2, sfs, FUN = &#39;/&#39;) %&gt;% head(10),
  main=&quot;sequencing depth&quot;,
  cluster_cols=FALSE,
  cluster_rows=FALSE,
  annotation=ann
)</code></pre>
<p><img src="day02_files/figure-html/heatmap%20sequencing_depth-1.png" width="672" /></p>
<pre class="r"><code>ggplot(ann) +
  geom_point(aes(x=sfs, sizeFactor)) +
  geom_abline(intercept = 0, slope = 1, color=&quot;red&quot;) +
  theme_minimal() +
  xlab(&quot;Size factor sequencing depth&quot;) +
  ylab(&quot;Median of ratios sizefactor&quot;)</code></pre>
<p><img src="day02_files/figure-html/seq%20depth%20factors%20vs%20median%20of%20ratios%20factors-1.png" width="672" /></p>
</div>
<div id="b.-dispersions" class="section level3">
<h3>b. Dispersions</h3>
<p>Why do we need to model count data ?</p>
<pre class="r"><code>ggplot(data=counts(dds_clean) %&gt;% data.frame()) +
  geom_histogram(
    aes(x=BM_CTRL_2),
    stat=&#39;bin&#39;, bins=200
  ) +
  theme_minimal()</code></pre>
<p><img src="day02_files/figure-html/plotcounts-1.png" width="672" /></p>
<p>Recall that:</p>
<ul>
<li>Most genes have no counts (0-inflation)</li>
<li>very large range of values</li>
</ul>
<p>Even a non-trained eye can see that this is not normally distributed,
so what distribution do we use ? For count data in general:</p>
<ul>
<li>Binomial distribution: e.g. number of heads upon tossing a coin n
times.</li>
<li>Poisson distribution: ~= binomial though n = infinite and event is
very small (e.g. lottery). Appropriate for data where mean ==
variance</li>
</ul>
<p>In RNA-seq data we have a very large number of RNA molecules, finding
a particular gene ~= winning the lottery ?</p>
<p>Let’s have a look at the mean and variance of our counts within TG
treated BM cells (the replicates).</p>
<p><img src="day02_files/figure-html/mean%20vs%20variance-1.png" width="672" /></p>
<blockquote>
<p><strong>Poll 2.3 </strong>: what kind of distribution would be the
most appropriate for this data ?</p>
</blockquote>
<p>A few remarks:</p>
<ul>
<li>Mean != variance</li>
<li>higher expressed genes tend to have higher variances</li>
<li>the variance in the ‘lower expressed regions’ is quite
scattered</li>
</ul>
<p>“heteroscedasticity”</p>
<p>If we would have money (and time) for a very large number of
replicates, the high expression data points would move closer to the red
line -&gt; Poisson.</p>
<p>In practice, we stick to the negative binomial distribution as it’s a
good approximation for data where mean &lt; variance</p>
<p>Enter: <strong>Dispersion</strong></p>
<div id="gene-wise-dispersion-estimates" class="section level4">
<h4>gene-wise dispersion estimates</h4>
<p>Within group variability (e.g. between replicates) is modeled by
parameter <code>α</code>, which describes variance of counts with:
<code>Var = μ + α*μ²</code>, where <code>α</code> = dispersion,
<code>Var</code> = variance, and <code>μ</code> = mean. Accurate
estimation of <code>α</code> is critical:</p>
<ul>
<li>not a problem for large studies with many replicates</li>
<li>but usually we only have 2-3 replicates =&gt; estimates of
<code>α</code> per gene are usually unreliable (genes with similar means
usually have different dispersion values).</li>
</ul>
<p>What do the dispersions represent ?</p>
<ul>
<li>dispersions are inversely related to the mean</li>
<li>directly related to the variance</li>
</ul>
<p>(Higher for small mean counts, lower for high mean counts)</p>
</div>
<div id="mean-dispersion-relationship-shrinking" class="section level4">
<h4>mean-dispersion relationship &amp; shrinking</h4>
<p>DESeq2 shares dispersion information across genes to better estimate
variation. The assumption here is that genes with <em>similar expression
values</em> have <em>similar dispersion values</em>.</p>
<pre class="r"><code>plotDispEsts(dds_clean)</code></pre>
<p><img src="day02_files/figure-html/plot%20dispersions-1.png" width="672" /></p>
<p>what would be ‘bad’ dispersion ?</p>
<pre class="r"><code>metadata_thrash &lt;- metadata 
metadata_thrash$celltype &lt;- sample(metadata$celltype)
metadata_thrash$condition &lt;- sample(metadata$condition)
dds_thrash &lt;- DESeqDataSetFromMatrix(countData=data, colData=metadata_thrash, design=~celltype+condition+celltype:condition)
dds_thrash &lt;- DESeq(dds_thrash)</code></pre>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<pre class="r"><code>plotDispEsts(dds_thrash)</code></pre>
<p><img src="day02_files/figure-html/baddisp-1.png" width="672" /></p>
<p>Signs of a ‘good’ dispersion plot:</p>
<ul>
<li>data should scatter around the curve</li>
<li>dispersion should decrease with increasing mean of counts</li>
<li>‘clouds’ or ‘wonky shapes’ often indicate contamination /
outliers</li>
</ul>
</div>
</div>
<div id="c.-fitting-model-and-testing" class="section level3">
<h3>c. fitting model and testing</h3>
<p>Finally we get a fit for each gene, and a Wald test can be used to
test for significance of coefficients of interest.</p>
<p>Luckily DESeq2 wraps this in a convenient function as well:
<em>results()</em>. By default we obtain DE genes for the last
coefficient, though you can specify contrasts to get your comparison of
interest (see tomorrow).</p>
<blockquote>
<p><strong>Task</strong> Please run results() on the dds_clean object
and have a look at the result.</p>
</blockquote>
<pre class="r"><code>dds_res &lt;- results(dds_clean)
dds_res</code></pre>
<pre><code>## log2 fold change (MLE): condition TG vs DMSO 
## Wald test p-value: condition TG vs DMSO 
## DataFrame with 27351 rows and 6 columns
##                   baseMean log2FoldChange     lfcSE      stat      pvalue
##                  &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;
## ENSG00000000419 2285.01560       0.317449 0.1790600  1.772863 7.62514e-02
## ENSG00000000457  775.42002       0.334962 0.0873551  3.834485 1.25827e-04
## ENSG00000000460 1300.43696       0.963262 0.0939353 10.254526 1.12928e-24
## ENSG00000000938    2.45443      -0.325046 0.7933991 -0.409688 6.82035e-01
## ENSG00000000971 1626.92877       0.309393 0.1271822  2.432678 1.49876e-02
## ...                    ...            ...       ...       ...         ...
## ENSG00000273486  40.217168      0.0672792  0.204936  0.328295    0.742689
## ENSG00000273487   0.162242      0.5041292  3.950877  0.127599    0.898466
## ENSG00000273488  13.780769      0.4674108  0.375660  1.244237    0.213412
## ENSG00000273489   1.064893     -1.1261042  1.341504 -0.839434    0.401226
## ENSG00000273492   0.856644      0.9590323  1.652038  0.580515    0.561568
##                        padj
##                   &lt;numeric&gt;
## ENSG00000000419 1.59065e-01
## ENSG00000000457 5.89362e-04
## ENSG00000000460 2.57159e-23
## ENSG00000000938 7.98272e-01
## ENSG00000000971 4.10219e-02
## ...                     ...
## ENSG00000273486    0.840324
## ENSG00000273487          NA
## ENSG00000273488    0.356482
## ENSG00000273489    0.565374
## ENSG00000273492    0.707109</code></pre>
<ul>
<li>baseMean: mean of normalized counts for all samples</li>
<li><strong>log2FoldChange: log2 fold change</strong></li>
<li>lfcSE: standard error</li>
<li>stat: Wald statistic</li>
<li>pvalue: Wald test p-value</li>
<li><strong>padj: BH adjusted p-values</strong></li>
</ul>
<blockquote>
<p><strong>Poll 2.4:</strong> What does a positive log2 foldChange and a
significant padj value reflect ?</p>
</blockquote>
<ul>
<li>how many genes do we get with padj &lt; 0.05?</li>
</ul>
<pre class="r"><code>dds_res %&gt;% subset(padj &lt; 0.05) %&gt;% nrow()</code></pre>
<pre><code>## [1] 8532</code></pre>
<ul>
<li>get the 20 significantly expressed genes with the highest
log2FoldChange.</li>
</ul>
<pre class="r"><code>dds_res %&gt;% 
  data.frame() %&gt;% 
  subset(padj &lt; 0.05) %&gt;% 
  arrange(abs(log2FoldChange), descending=TRUE) %&gt;% head(20)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["baseMean"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["log2FoldChange"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["lfcSE"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["stat"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["pvalue"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["padj"],"name":[6],"type":["dbl"],"align":["right"]}],"data":[{"1":"3790.108","2":"-0.1169817","3":"0.04563335","4":"-2.563513","5":"0.010361893","6":"0.029868034","_rn_":"ENSG00000144579"},{"1":"1980.431","2":"-0.1225237","3":"0.03846077","4":"-3.185680","5":"0.001444143","6":"0.005304359","_rn_":"ENSG00000114125"},{"1":"3277.863","2":"-0.1242576","3":"0.05213396","4":"-2.383429","5":"0.017152178","6":"0.046008939","_rn_":"ENSG00000118900"},{"1":"14936.948","2":"-0.1276037","3":"0.05209341","4":"-2.449517","5":"0.014304811","6":"0.039363087","_rn_":"ENSG00000122565"},{"1":"3170.810","2":"-0.1279751","3":"0.04920218","4":"-2.601005","5":"0.009295104","6":"0.027220686","_rn_":"ENSG00000223380"},{"1":"4694.747","2":"-0.1301570","3":"0.05270412","4":"-2.469578","5":"0.013527234","6":"0.037497739","_rn_":"ENSG00000168175"},{"1":"3840.813","2":"0.1328259","3":"0.05611064","4":"2.367214","5":"0.017922571","6":"0.047743321","_rn_":"ENSG00000100568"},{"1":"3160.154","2":"0.1338326","3":"0.05504994","4":"2.431113","5":"0.015052508","6":"0.041164559","_rn_":"ENSG00000108443"},{"1":"1600.127","2":"0.1355010","3":"0.05397121","4":"2.510616","5":"0.012052082","6":"0.034005919","_rn_":"ENSG00000174227"},{"1":"2092.895","2":"0.1393615","3":"0.05729098","4":"2.432520","5":"0.014994144","6":"0.041029810","_rn_":"ENSG00000061936"},{"1":"2226.112","2":"-0.1424242","3":"0.05893415","4":"-2.416667","5":"0.015663356","6":"0.042582215","_rn_":"ENSG00000084112"},{"1":"7396.732","2":"-0.1433622","3":"0.05843881","4":"-2.453201","5":"0.014159107","6":"0.039032256","_rn_":"ENSG00000005844"},{"1":"1373.535","2":"0.1436665","3":"0.05855470","4":"2.453543","5":"0.014145655","6":"0.039001223","_rn_":"ENSG00000105875"},{"1":"3070.223","2":"0.1448491","3":"0.06021732","4":"2.405439","5":"0.016153041","6":"0.043755333","_rn_":"ENSG00000172943"},{"1":"2260.323","2":"-0.1465682","3":"0.05499220","4":"-2.665254","5":"0.007693022","6":"0.023147791","_rn_":"ENSG00000179933"},{"1":"3790.109","2":"-0.1473921","3":"0.05275600","4":"-2.793846","5":"0.005208529","6":"0.016494650","_rn_":"ENSG00000103018"},{"1":"2266.977","2":"-0.1494905","3":"0.06285824","4":"-2.378217","5":"0.017396596","6":"0.046542854","_rn_":"ENSG00000157014"},{"1":"12880.910","2":"-0.1536982","3":"0.06026682","4":"-2.550295","5":"0.010763189","6":"0.030859216","_rn_":"ENSG00000120948"},{"1":"6005.095","2":"-0.1547289","3":"0.06350332","4":"-2.436549","5":"0.014828163","6":"0.040624883","_rn_":"ENSG00000149187"},{"1":"2621.111","2":"-0.1553704","3":"0.05656124","4":"-2.746942","5":"0.006015380","6":"0.018690497","_rn_":"ENSG00000147789"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="d.-exporting-results-and-further-analyses"
class="section level3">
<h3>d. Exporting results and further analyses</h3>
<div class="figure">
<img src="images/day2_analysis_tools.png" style="width:80.0%" alt="" />
<p class="caption">Export data</p>
</div>
<blockquote>
<p><strong>Task</strong>: Please export your results table to a csv /
tsv file.</p>
</blockquote>
<p>hint: write_tsv, write_csv(), write.table(), write_delim() ,…</p>
<pre class="r"><code>write.table(dds_res %&gt;% data.frame, file=&quot;DMSO_vs_TG_BMcells.tsv&quot;, col.names=T, row.names = T, sep = &quot;\t&quot;, quote=F)</code></pre>
<p>shiny apps on MPI-IE website (<a
href="https://rsconnect.ie-freiburg.mpg.de/shiny/"
class="uri">https://rsconnect.ie-freiburg.mpg.de/shiny/</a>)</p>
<div class="figure">
<img src="images/day2_shinyapps.png" style="width:80.0%" alt="" />
<p class="caption">Shiny apps</p>
</div>
<p><em>Genes2Functions</em></p>
<div class="figure">
<img src="images/day2_genes2functions_maplot.png" style="width:80.0%"
alt="" />
<p class="caption">genes2functions_maplot</p>
</div>
<div class="figure">
<img src="images/day2_genes2functions_computing.png" style="width:80.0%"
alt="" />
<p class="caption">genes2functions_GO</p>
</div>
<div class="figure">
<img src="images/day2_genes2functions_dotplot.png" style="width:80.0%"
alt="" />
<p class="caption">genes2functions_GO</p>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
