<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Bioinfo-Core @ MPI-IE" />


<title>DESeq2 Analysis with R: outline</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Rdeseq2 - RNA-seq analysis</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    
  </a>
</li>
<li>
  <a href="outline.html">
    <span class="fa fa-check"></span>
     
    outline
  </a>
</li>
<li>
  <a href="day01.html">
    <span class="fa fa-battery-quarter"></span>
     
    Day 1
  </a>
</li>
<li>
  <a href="day02.html">
    <span class="fa fa-battery-half"></span>
     
    Day 2
  </a>
</li>
<li>
  <a href="day03.html">
    <span class="fa fa-battery-three-quarters"></span>
     
    Day 3
  </a>
</li>
<li>
  <a href="day04.html">
    <span class="fa fa-battery-full"></span>
     
    Day 4
  </a>
</li>
<li>
  <a href="summary.html">
    <span class="fa fa-truck-medical"></span>
     
    summary
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">DESeq2 Analysis with R: outline</h1>
<h4 class="author">Bioinfo-Core @ MPI-IE</h4>
<h4 class="date">Fri Apr 28 09:49:06 2023</h4>

</div>


<hr />
<div id="libraries" class="section level1">
<h1>Libraries</h1>
<p>We will use a couple of libraries in this course. We have to make
sure these are installed. These installations can take a while. So itâ€™s
best to tackle this first. The following code chunk will checks if a
required library is available and will attempt to install it if this is
not the case.</p>
<pre class="r"><code># Bioconductor
if (!require(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
if (!require(&quot;DESeq2&quot;, quietly = TRUE))
    BiocManager::install(&quot;DESeq2&quot;)
if (!require(&quot;EnsDb.Hsapiens.v75&quot;, quietly = TRUE))
    BiocManager::install(&quot;EnsDb.Hsapiens.v75&quot;)

# CRAN
if (!require(&quot;tidyverse&quot;, quietly = TRUE))
    install.packages(&quot;tidyverse&quot;)
if (!require(&quot;pheatmap&quot;, quietly = TRUE))
    install.packages(&quot;pheatmap&quot;)
if (!require(&quot;ggrepel&quot;, quietly = TRUE))
    install.packages(&quot;ggrepel&quot;)
if (!require(&quot;UpSetR&quot;, quietly = TRUE))
    install.packages(&quot;UpSetR&quot;)
if (!require(&quot;ashr&quot;, quietly = TRUE))
    install.packages(&quot;ashr&quot;)</code></pre>
<p>After running the above chunk, make sure the libraries can be
loaded:</p>
<pre class="r"><code>library(&quot;DESeq2&quot;)
library(&quot;tidyverse&quot;)
library(&quot;pheatmap&quot;)
library(&quot;ggrepel&quot;)
library(&quot;EnsDb.Hsapiens.v75&quot;)
library(&quot;UpSetR&quot;)
library(&quot;ashr&quot;)</code></pre>
<hr />
</div>
<div id="what-you-will-hopefully-learn" class="section level1">
<h1>What you will (hopefully) learn</h1>
<p>After the course you will be able to</p>
<ul>
<li>Perform a full bioinformatics workflow (DESeq2): data input <span
class="math inline">\(\to\)</span> output<br />
</li>
<li>Understand the theory behind RNA-seq data generation and
analysis</li>
<li>Perform exploratory analysis: quality controls, filters<br />
</li>
<li>Visualize data and results with modern tools<br />
</li>
<li>Apply complex (multifactorial) experimental designs<br />
</li>
<li>Obtain and visualise differential genes from simple or complex
comparison setups<br />
</li>
<li>Share workflows with your colleagues.</li>
</ul>
<hr />
</div>
<div id="credits" class="section level1">
<h1>Credits</h1>
<p>There is lots of useful material out there. This course is heavily
influenced by those tutorials:</p>
<ul>
<li><a
href="https://hbctraining.github.io/DGE_workshop_salmon_online/">Harvard
Chan bioinformatics coreâ€™s DGE workshop using Salmon counts</a></li>
<li><a
href="http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DESeq2
vignette</a></li>
<li><a
href="https://github.com/tavareshugo/tutorial_DESeq2_contrasts">Hugo
Tavaresâ€™ contrast tutorial</a></li>
</ul>
<p>We will also be using publicly available data, which has been altered
for educational purposes (and should thus be treated as such).</p>
<hr />
</div>
<div id="overview" class="section level1">
<h1>Overview</h1>
<div id="day-i-getting-and-understanding-the-data"
class="section level2">
<h2>1. Day I: Getting and understanding the data</h2>
<ul>
<li>Warmup &amp; R-epetition</li>
<li>Data, Metadata &amp; Design</li>
<li>Data Exploration</li>
<li>Quality Control</li>
</ul>
</div>
<div id="day-ii-data-transformations-and-model" class="section level2">
<h2>2. Day II: Data Transformations and Model</h2>
<ul>
<li>Quality Controls &amp; Filtering</li>
<li>Data Transformations and Normalization</li>
<li>Size Factors</li>
<li>Modeling Count Data: NB and Over-dispersion</li>
<li>Running DESeq2</li>
</ul>
</div>
<div id="day-iii-differentially-expressed-genes" class="section level2">
<h2>3. Day III: Differentially Expressed Genes</h2>
<ul>
<li>Multifactorial Designs &amp; Complex Contrasts</li>
<li>Hypothesis tests: Wald test &amp; LRT</li>
<li>Inspection &amp; Visualization</li>
<li>Gene Annotations</li>
<li>Exporting results</li>
</ul>
</div>
<div id="day-iv-on-your-own" class="section level2">
<h2>4. Day IV: On your own</h2>
<ul>
<li>Run a multifactorial DESeq2 Analysis. Weâ€™ll be around to help.</li>
</ul>
<hr />
</div>
</div>
<div id="recap" class="section level1">
<h1>Recap</h1>
<p>The Rstudio GUI.</p>
<div class="figure">
<img src="images/outline_script.png" alt="" />
<p class="caption">An Rscript in the Rstudio GUI</p>
</div>
<div class="figure">
<img src="images/outline_book.png" alt="" />
<p class="caption">An R notebook in the Rstudio GUI</p>
</div>
<blockquote>
<p><strong>Poll 0.1</strong>: How will you keep track of the code,
course material and notes during this course ?</p>
</blockquote>
<hr />
</div>
<div id="set-up-our-project" class="section level1">
<h1>Set up our project</h1>
<blockquote>
<p><strong>Task</strong>: Create a new project, cloning the datafiles we
will be using from github.</p>
</blockquote>
<p>If you are working in workbench, start a new session.</p>
<p>In your session (local or workbench):</p>
<ul>
<li>file -&gt; new project -&gt; â€˜Version Controlâ€™ -&gt; git</li>
</ul>
<blockquote>
<p>Repository URL: <a href="https://github.com/maxplanck-ie/Rdeseq2"
class="uri">https://github.com/maxplanck-ie/Rdeseq2</a><br />
Project directory name: chose a folder name</p>
</blockquote>
<p>! For workbench users:</p>
<ul>
<li>create the project as a subdirectory under your group volume
(e.g.Â /data/manke/group/deboutte), or a processing volume
(/data/processing/deboutte)</li>
<li>Remember to choose version 4.1.3</li>
</ul>
<p>If it worked, you should have a new session, and in your console you
will retrieve your project directory:</p>
<blockquote>
<p>getwd()</p>
</blockquote>
<p>and you should be able to see the data folder:</p>
<blockquote>
<p>list.files(â€˜dataâ€™)</p>
</blockquote>
<p>which should contain these folders:</p>
<ul>
<li>genesets</li>
<li>mpp</li>
<li>myeloma</li>
</ul>
<p>How it should look like:</p>
<div class="figure">
<img src="images/outline_clonedrepo.png" alt="" />
<p class="caption">An new R notebook in our freshly cloned project</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
